stack/framework/hal/chips/cortus/cortus_counter.c

```diff
@@ -0,0 +1,106 @@
+/** OSS-7 - An opensource implementation of the DASH7 Alliance Protocol for ultra
+ * lowpower wireless sensor communication
+ *
+ * Copyright 2018 University of Antwerp
+ * Copyright 2018 Cortus
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.* OSS-7 - An opensource implementation of the DASH7 Alliance Protocol for ultra
+ * lowpower wireless sensor communication
+ *
+ * \file cortus_counter.c
+ *
+ * \author yousra.amrani@cortus.com
+ *
+ */
+
+
+#include <stdbool.h>
+#include <stdint.h>
+#include "errors.h"
+#include "hwcounter.h"
+
+#include "machine/counter.h"
+
+#define HWCOUNTER_NUM 2
+#define FREQ_CPU 160 // 16*10e6
+
+static uint32_t start_value_counter1;
+static uint32_t start_value_counter2;
+
+error_t hw_counter_init(hwcounter_id_t counter_id, uint16_t time_usecond, bool repeat)
+{
+   if(counter_id >= HWCOUNTER_NUM)
+      return ESIZE;
+
+   if (counter_id == 1)
+   {
+       start_value_counter1 = (time_usecond * FREQ_CPU) / 10;
+       if ( repeat )
+           counter1->reload = start_value_counter1;
+       else
+           counter1->reload = 0x00;
+       counter1->value = start_value_counter1;
+       counter1->expired = false;
+   }
+   else
+   {
+       start_value_counter2 = (time_usecond * FREQ_CPU) / 10;
+       if ( repeat )
+           counter2->reload = start_value_counter2;
+       else
+           counter2->reload = 0x00;
+       counter2->value = start_value_counter2;
+       counter2->expired = false;
+   }
+   return SUCCESS;
+}
+
+hwcounter_tick_t hw_counter_getvalue(hwcounter_id_t counter_id)
+{
+   if(counter_id >= HWCOUNTER_NUM )
+      return 0;
+
+   if (counter_id == 1)
+       return (counter1->value);
+   else
+       return(counter2->value);
+}
+
+uint16_t hw_counter_get_elapsed_time(hwcounter_id_t counter_id)
+{
+    if(counter_id >= HWCOUNTER_NUM )
+        return 0;
+
+    uint16_t value;
+
+    if (counter_id == 1)
+    {
+        value = (uint16_t)( ((start_value_counter1 - counter1->value) * 10) / FREQ_CPU);
+    }
+    else
+    {
+    	value = (uint16_t)( ((start_value_counter2 - counter2->value) * 10) / FREQ_CPU);
+    }
+    return value;
+}
+
+bool hw_counter_is_expired(hwcounter_id_t counter_id)
+{
+    if(counter_id >= HWCOUNTER_NUM )
+        return false;
+
+    if (counter_id == 1)
+        return (counter1->expired);
```

Comment:
where is the expired flag set?

Reply:
The flag is set by the hardware directly.